python/dgl/convert.py

```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
+        u, v, urange, vrange = utils.graphdata2tensors(data, idtype, bipartite=True)
+        node_tensor_dict[(sty, ety, dty)] = (u, v)
+        if need_infer:
+            num_srcnodes_dict[sty] = max(num_srcnodes_dict[sty], urange)
+            num_dstnodes_dict[dty] = max(num_dstnodes_dict[dty], vrange)
+        else:  # sanity check
+            if num_srcnodes_dict[sty] < urange:
+                raise DGLError('The given number of nodes of node type {} must be larger than'
+                               ' the max ID in the data, but got {} and {}.'.format(
+                                   sty, num_srcnodes_dict[sty], urange - 1))
+            if num_dstnodes_dict[dty] < vrange:
+                raise DGLError('The given number of nodes of node type {} must be larger than'
+                               ' the max ID in the data, but got {} and {}.'.format(
+                                   dty, num_dstnodes_dict[dty], vrange - 1))
+    # Create the graph
+
+    # Sort the ntypes and relation tuples to have a deterministic order for the same set
+    # of type names.
+    srctypes = list(sorted(num_srcnodes_dict.keys()))
+    dsttypes = list(sorted(num_dstnodes_dict.keys()))
+    relations = list(sorted(node_tensor_dict.keys()))
+
+    num_nodes_per_type = utils.toindex(
+        [num_srcnodes_dict[ntype] for ntype in srctypes] +
+        [num_dstnodes_dict[ntype] for ntype in dsttypes], "int64")
+    srctype_dict = {ntype: i for i, ntype in enumerate(srctypes)}
+    dsttype_dict = {ntype: i + len(srctypes) for i, ntype in enumerate(dsttypes)}
+
+    meta_edges_src = []
+    meta_edges_dst = []
+    etypes = []
+    rel_graphs = []
+    for srctype, etype, dsttype in relations:
+        meta_edges_src.append(srctype_dict[srctype])
+        meta_edges_dst.append(dsttype_dict[dsttype])
+        etypes.append(etype)
+        src, dst = node_tensor_dict[(srctype, etype, dsttype)]
+        g = create_from_edges(src, dst, 'SRC/' + srctype, etype, 'DST/' + dsttype,
+                              num_srcnodes_dict[srctype], num_dstnodes_dict[dsttype])
+        rel_graphs.append(g)
+
+    # metagraph is DGLGraph, currently still using int64 as index dtype
+    metagraph = graph_index.from_coo(
+        len(srctypes) + len(dsttypes), meta_edges_src, meta_edges_dst, True)
+    # create graph index
+    hgidx = heterograph_index.create_heterograph_from_relations(
+        metagraph, [rgrh._graph for rgrh in rel_graphs], num_nodes_per_type)
+    retg = DGLBlock(hgidx, (srctypes, dsttypes), etypes)
+
+    return retg.to(device)
+
+def block_to_graph(block):
+    """Convert a :class:`DGLBlock` object to a :class:`DGLGraph`.
+
+    DGL will rename all the input node types by suffixing with ``_src``, and
+    all the output node types by suffixing with ``_dst``.
+
+    The storage of the graph will be shared with that of the input block.
+
+    Features on the returned graph will be preserved, shared with that of the input block.
```

Comment:
What's the motivation for sharing storage and features? I feel we are generally not sharing storage and features across functions.

Reply:
I checked that `node_type_subgraph` and `edge_type_subgraph` actually shared features while `edge_subgraph` does not share node features when `preserve_nodes=True`.  I think sharing features is cheap and we do have two practices so I'm not sure what to do here.  Looping in @jermainewang .

Reply:
I think Murphy's question is about whether it is a sharing or a shallow copy? Will mutation on the features of the result graph be visible to the input graph? In general, I would prefer not share the frame but let the columns refer to the same tensor storage. This is consistent with the transform operators such as [`to_simple`](https://docs.dgl.ai/generated/dgl.to_simple.html#dgl.to_simple).

Reply:
Yeah that's how this function is implemented.  Frames are brand new but the columns share the same storage.